### ISSUE TITLE
Cache external requests for integration tests

### ISSUE THREAD

**Issue by Panaetius** [SwissDataScienceCenter/renku-python#2506]
To speed up our tests (and not hammer external services as much) we should record web requests to be replayed in tests.

We could use something like https://pypi.org/project/pytest-recording/ which allows annotating tests with `@pytest.mark.vcr`.

Then we need to add a github actions cache (akin to https://github.com/SwissDataScienceCenter/renku-python/blob/master/.github/workflows/test_deploy.yml#L84 ) for our integration tests. 
The cache key should include:
- The current week
- the hash of the `tests/` directory
- a flag set at the top of the tests to manually trigger a cache miss
- (optionally)the hash of pyproject.toml. Not sure if this is beneficial (like it requests library changes)

Here we should call pytest with `--record-mode=once` if the cache was hit and `--record-mode=all` if the cache was a miss.

This should significantly speed up tests and prevent us from being rate-limited.

**Comment by mohammad-sdsc** [comment]
See also https://betamax.readthedocs.io/en/latest/integrations.html

**Comment by Panaetius** [comment]
I looked at betamax when creating this issue. Looking at their docs, the main differences to pytest-recording(pr) that I saw were:
- Need to manually patch requests `session` (pr monkeypatches it automatically)
- Need to manually specify a casette to use (pr automatically creates them based on the file path of a test)
- Config in code, no pytest level flags (pr allows passing CLI flags to force recreate recordings on a regular basis)

So its integration with pytest seems strictly worse for our use case, e.g. https://betamax.readthedocs.io/en/latest/usage_patterns.html#an-example-from-github3-py needs some config to enable things that pr supports out of the box.

I saw you use pytest-recording in the linked PR, so I guess you came to the same conclusion :slightly_smiling_face: Just wanted to add my reasoning for completeness' sake
